doc/configuring.md
# Configuring

All configuration options can be set in the `config` file.
See [config.sample](/examples/config.sample) for a commented example.
If run without a config file, the example file will be copied in place and
left behind for the user to edit.

The configuration file along with other data will be stored in a single
directory, `~/.lighter` by default (use `--lighterdir` option to override).
Paths specified in the configuration file are relative to this directory.

#### Bare-minimum
For Lighter to work, you will need to set the following options:
* `implementation`
* `server_key` and `server_crt` (unless `insecure_connection` is `1`)
and then configure the section for the chosen implementation.

You can request TLS certificates from CAs (e.g. letsencrypt) or you can
generate your own. The certificate must include the host names or IP addresses
you will use to contact Lighter or certificate validation will fail (in at
least some cases).

Example one-liner to generate a self-signed TLS certificate <sup>1</sup>:
```
openssl req -newkey rsa:2048 -nodes -keyout server.key -x509 -days 365 -out server.crt -subj "/CN=node.example.com" -extensions SAN -config <(cat /etc/ssl/openssl.cnf <(printf "\n[SAN]\nsubjectAltName=DNS:node.example.com,DNS:lighter,DNS:localhost,IP:127.0.0.1,IP:::1"))
```



### lighter section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `implementation` <sup>2</sup> | Implementation to use (possible values: `clightning`, `eclair`, `electrum`, `lnd`; no default) |
| `insecure_connection`         | Set to `1` to make Lighter listen in cleartext (default `0`). Implies disabling macaroons. |
| `port`                        | Lighter's listening port (default `1708`)                                  |
| `server_key`                  | Private key path (default `./certs/server.key`)                            |
| `server_crt`                  | Certificate (chain) path (default `./certs/server.crt`)                    |
| `logs_dir`                    | Location <sup>4</sup> to hold log files (default `./logs`)                 |
| `logs_level`                  | Desired console log level (possible values: `critical`, `error`, `warning`, `info`, `debug`; default `info`) |
| `db_dir`                      | Location to hold the database (default `./db`)                             |
| `macaroons_dir`               | Location to hold macaroons (default `./macaroons`)                         |
| `disable_macaroons` <sup>3</sup> | Set to `1` to disable macaroons authentication (default `0`)            |

### cliter section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `rpcserver`                   | Lighter host[:port] (default `localhost:1708`)                             |
| `tlscert`                     | Lighter certificate (chain) path (default `./certs/server.crt`)          |
| `macaroon`                    | Lighter macaroon path (default `./macaroons/admin.macaroon`)             |
| `insecure`                    | Set to `1` to connect to Lighter in cleartext  (default `0`)               |
| `no_macaroon`                 | Set to `1` to connect to Lighter with no macaroon (default `0`)            |


### clightning section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `cl_rpc_dir` <sup>5</sup>     | Location <sup>4</sup> containing `cl_rpc`                                  |
| `cl_rpc` <sup>6</sup>         | JSON-RPC socket name (default `lightning-rpc`)                             |

### eclair section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `ecl_host`                    | Host <sup>7</sup> (default `localhost`)                                    |
| `ecl_port` <sup>8</sup>       | Port (default `8080`)                                                      |

### electrum section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `ele_host`                    | Host <sup>7</sup> (default `localhost`)                                    |
| `ele_port`                    | Port (default `7777`)                                                      |
| `ele_user`                    | User (default `user`)                                                      |

### lnd section

| Variable                      | Description                                                                |
| ----------------------------- | -------------------------------------------------------------------------- |
| `lnd_host`                    | Host <sup>7</sup> (default `localhost`)                                    |
| `lnd_port`                    | Port (default `10009`)                                                     |
| `lnd_cert_dir`                | Location <sup>4</sup> containing `lnd_cert`                                |
| `lnd_cert`                    | TLS certificate name (default `tls.cert`)                                  |

#### Notes

1. _requires the package_ `openssl` _to be installed, which provides_
   `/etc/ssl/openssl.cnf`
2. _implementation value is case-insensitive_
3. _running Lighter on mainnet with macaroons disabled has severe security
   implications and is highly discouraged, don't do this unless you know
   what you're doing_
4. _location must be a directory;
path can be absolute or relative (to Lighter's data directory)_
5. _usually ~/.lightning/\<network\>_
6. _option:_ `rpc-file` _the JSON-RPC socket needs to be owned by the same user
   Lighter is running as_
7. _host can be an IP or a FQDN_
8. _option:_ `eclair.api.port` _(usually 8080)_


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 1.2.0 - 2019-11-22

### Added
- default value for `expiry_time` in `CreateInvoice`
- paring procedure (`make pairing`)
- systemd example service
- `sqlalchemy` ORM for better DB handling and `alembic` as DB migration tool
- support for latest implementation versions
(c-lightning v0.7.3, eclair v0.3.2, lnd v0.8.0-beta)
- proto: added `state` field to `CheckInvoice` message
- proto: added `UnlockNode` API
- proto: added `unlock_node` field to `UnlockLighter` message
- secure: added non-interactive mode

### Changed
- default value of config var `DOCKER`
- improved security documentation
- updated pips and docker base image
- errors: changed double quotes into single around `%PARAM%`
- secure: ask for implemenetation password 2 times

### Deprecated
- docker assisted usage

### Fixed
- cliter: fixed bad escaping of double quotation mark
- Makefile: fixed pips install calling secure target
- secure: fixed password mismatch on new db case


## 1.1.1 - 2019-10-02

### Added
- cliter: add check for incompatible options
- cliter: add `--insecure`, `--macaroon`, `--no-macaroon`, `--rpcserver`,
`--tlscert`, `--version` options
- cliter: add installation in docker image
- support for c-lightning v0.7.2 (v0.7.2.1)

### Changed
- c-lightning: changed `listpayments` (deprecated) to `listsendpays`
- cliter: changed output format to JSON
- cliter: cli errors are now redirected to stderr (except for lighter errors)
and exit code is 1
- cliter: lighter errors are now printed in JSON format (`code` and `details`)
- removed default description from invoices

### Fixed
- c-lightning: fixed parsing of fallback address in `DecodeInvoice`
- c-lightning: fixed `PayOnChain`
- cliter: fixed packaging (add generation of proto files in `setup.py`)
- lnd: fixed closing txid decoding


## 1.1.0 - 2019-08-21

### Added
- support for latest implementation versions
(c-lightning v0.7.1, eclair v0.3.1, lnd v0.7.1-beta)
- error log when a request is aborted
- new common error mappings: `closechannel_failed`, `payinvoice_failed` and
`payinvoice_pending`
- dynamic implementation timeout based on client one
- log message when Lighter is locked
- config: added option to change console logging level
- eclair: added `OpenChannel`
- proto: added `amount_received_bits` field to `Invoice` message
- proto: added `balance_confirmed` field to `WalletBalanceResponse` message
- proto: added `CloseChannel` API (full support)
- proto: added `color` field to `Peer` message
- proto: added `(local|remote)_reserve_sat` fields to `Channel` message
- proto: added `LockLighter` API
- proto: added more granular balances to `ChannelBalanceResponse`
- proto: added `node_uri` field to `GetInfoResponse` message
- proto: added `payment_request` field to `Invoice` message
- proto: added `private` field to `Channel` message
- proto: added `state` and `active` fields to `Channel` message

### Changed
- added internal bech32 amount decode to avoid calling the node
(currently lnd's decode doesn't handle amounts below 1 sat)
- added operations to invoices macaroon
- code of conduct
- description of `unimplemented_method` and `unimplemented_parameter` errors
- improved implementation errors mapping
- made `make secure` mandatory
- refactored `Crypter` and `DbHandler`
- cli: increased timeout + log fix
- cli: renamed from `lit-cli` to `cliter`

### Fixed
- minor bugs in `PayInvoice`, `ListChannels`, `OpenChannel`, `ListPeers` and
`GetInfo`
- secure: added keyboard interrupt catching and decryption error handling

### Security
- generation of macaroon rootkey at runtime
- password documentation and auto generation
- using different derived key for each secret
- using `/dev/random` by default to generate password and salts


## 1.0.0 - 2019-04-23

### Added
- added security support
- added a CLI (`lit-cli`) with bash completion
- Makefile: added `secure` and `cli` targets
- proto: added `UnlockerServicer`
- errors: added mappings
- config: added variables `DB_DIR`, `DB_NAME`, `DISABLE_MACAROONS`,
`INSECURE_CONNECTION` and `MACAROON_DIR`
- add operation logging

### Changed
- Makefile: improved `clean` target (now removes also autogenerated files)
- proto/HopHint: renamed `fee_base_bits` to `fee_base_msat`

### Removed
- config: removed variables `ALLOW_INSECURE_CONNECTION`,
`ALLOW_SECURE_CONNECTION`, `ECL_PASS` `LND_MACAROON` and `LND_MACAROON_DIR`

### Fixed
- unix_make: improved bash portability and arch detect for macOS
- utils: fixed `check_connection` (now async and persistent until successful)

### Security
- added macaroon support
- added a database to store implementation secrets and macaroon root key
- added `secure` process for handling secrets


## 0.2.2 - 2019-03-06

### Fixed

- proto: fixed OpenChannel rpc definition
- lnd: fixed missing `max_precision` in `funding_bits` conversion of OpenChannel
- `KeyboardInterrupt` was not triggering `_slow_exit`

## 0.2.1 - 2019-03-04

### Changed

- Updated supported implementation versions

### Fixed

- Utils module tests (oops)
- c-lightning: wrong parameter for timestamp


## 0.2.0 - 2019-02-12

### Added

- New APIs (see support table):
  * ListInvoices
  * ListPayments
  * ListTransactions
  * OpenChannel
  * PayOnChain
- Supported APIs table
- gRPC version update (1.18.0)
- Node.js proto compilation example
- New error mappings
- lnd connection handling decorator
- Checks for required parameters

### Fixed

- Makefile: targets lint and test work on first run
- Dockerfile generation on arm
- c-lightning invoice creation (`check_value` usage)
- lighter.proto reordering (alphabetical sort)


## 0.1.0 - 2018-12-04

### Added

- Protobuf definitions
- Lighter dispatcher
- Support for c-lightning node
- Support for eclair node
- Support for lnd node

lighter_bitcoin/settings.py
# Copyright (C) 2018 inbitcoin s.r.l.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published
# by the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""
Settings module for Lighter

WARNING: only Python's core module imports here, to avoid breaking build system
         and import loops
"""

from os import path
from pathlib import Path

# Empty variables are set at runtime
# Some variables contain default values, could be overwritten

IMPLEMENTATION = ''

L_DATA = path.join(str(Path.home()), '.lighter')
L_CONFIG = path.join(L_DATA, 'config')

PKG_NAME = 'lighter_bitcoin'
PIP_NAME = 'lighter-bitcoin'

HOST = '0.0.0.0'
PORT = '1708'
LIGHTER_ADDR = ''
INSECURE_CONNECTION = 0
SERVER_KEY = './certs/server.key'
SERVER_CRT = './certs/server.crt'
IMPLEMENTATION_SECRETS = False
IMPL_SEC_TYPE = ''

# Macaroons settings
RUNTIME_BAKER = None
DISABLE_MACAROONS = 0
MACAROONS_DIR = './macaroons'
MAC_ADMIN = 'admin.macaroon'
MAC_READONLY = 'readonly.macaroon'
MAC_INVOICES = 'invoices.macaroon'

# Security settings
MAC_ROOT_KEY = None
SALT_LEN = 32
ACCESS_TOKEN = b'lighter'
PASSWORD_LEN = 12
ENTROPY_BLOCKING = 1
SCRYPT_PARAMS = {
    'cost_factor': 2**15,
    'block_size_factor': 8,
    'parallelization_factor': 1,
    'key_len': 32
}

# lighter-secure settings
COLLECTING_INPUT = True
DONE = False
NEW_DB = False
SEARCHING_ENTROPY = True
FIRST_WORK_TIME = None
IDLE_COUNTER = 1
IDLE_MESSAGES = {
    1: {
        'msg': 'please keep generating entropy',
        'delay': 5
    },
    2: {
        'msg': 'more entropy, please',
        'delay': 15
    },
    3: {
        'msg': '...good things come to those who wait...',
        'delay': 30
    }
}

# DB settings
DB_DIR = './db'
DB_NAME = 'lighter.db'
DB_PATH = ''
ALEMBIC_CFG = path.join(path.dirname(__file__), 'migrations/alembic.ini')

# Server settings
GRPC_WORKERS = 10
GRPC_GRACE_TIME = 40
UNLOCKER_STOP = False
RUNTIME_STOP = False
RUNTIME_SERVER = None
THREADS = []

# cliter settings
RPCSERVER = 'localhost:1708'
TLSCERT = './certs/server.crt'
MACAROON = path.join(MACAROONS_DIR, MAC_ADMIN)
INSECURE = 0
NO_MACAROON = 0
CLI_RPCSERVER = None
CLI_TLSCERT = None
CLI_MACAROON = None
CLI_INSECURE = None
CLI_NO_MACAROON = None
CLI_TIMEOUT = 10
CLI_BASE_GRPC_CODE = 64

ENFORCE = True

TEST_HASH = '43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000'
MAIN_HASH = '6fe28c0ab6f1b372c1a6a246ae63f74f931e8365e15a089c68d6190000000000'

# RPC-based implementations settings
RPC_URL = ''
RPC_TRIES = 5
RPC_SLEEP = .5
RPC_CONN_TIMEOUT = 3.1

# c-lightning specific settings
CL_RPC = 'lightning-rpc'

# eclair specific settings
ECL_HOST = 'localhost'
ECL_PORT = 8080
ECL_PASS = ''

# electrum specific settings
ELE_HOST = 'localhost'
ELE_PORT = 7777
ELE_USER = 'user'
ELE_RELEASED_ADDRESSES = []

# lnd specific settings
LND_HOST = 'localhost'
LND_PORT = 10009
LND_CERT = 'tls.cert'
LND_ADDR = ''
LND_CREDS_SSL = ''
LND_CREDS_FULL = ''
LND_MAC = ''

# Common settings
IMPL_MIN_TIMEOUT = 2
IMPL_MAX_TIMEOUT = 180
RESPONSE_RESERVED_TIME = 0.3
THREAD_TIMEOUT = 3
CLOSE_TIMEOUT_NODE = 15
MAX_INVOICES = 200
INVOICES_TIMES = 3
EXPIRY_TIME = 420
DUST_LIMIT_SAT = 546

# Logging settings
LOGS_DIR = './logs'
LOGS_LIGHTER = 'lighter.log'
LOGS_MIGRATIONS = 'migrations.log'
LOG_TIMEFMT = '%Y-%m-%d %H:%M:%S %z'
LOG_TIMEFMT_SIMPLE = '%d %b %H:%M:%S'
LOGS_LEVEL = 'INFO'
LOG_LEVEL_FILE = 'DEBUG'
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'formatters': {
        'verbose': {
            'format':
            "[%(asctime)s] %(levelname).3s [%(name)s:%(lineno)s] %(message)s",
            'datefmt': LOG_TIMEFMT
        },
        'simple': {
            'format': '%(asctime)s %(levelname).3s: %(message)s',
            'datefmt': LOG_TIMEFMT_SIMPLE
        },
    },
    'handlers': {
        'console': {
            'level': LOGS_LEVEL,
            'class': 'logging.StreamHandler',
            'formatter': 'simple',
            'stream': 'ext://sys.stdout'
        },
    },
    'loggers': {
        '': {
            'handlers': ['console'],
            'level': 'DEBUG'
        },
    }
}
LOGGING_FILE = {
    'file': {
        'level': LOG_LEVEL_FILE,
        'class': 'logging.handlers.RotatingFileHandler',
        'filename': path.join(LOGS_DIR, LOGS_LIGHTER),
        'maxBytes': 1048576,
        'backupCount': 7,
        'formatter': 'verbose'
    }
}

ALL_PERMS = {
    '/lighter.Lightning/ChannelBalance': {
        'entity': 'balance',
        'action': 'read'
    },
    '/lighter.Lightning/CheckInvoice': {
        'entity': 'invoice',
        'action': 'read'
    },
    '/lighter.Lightning/CloseChannel': {
        'entity': 'channel',
        'action': 'write'
    },
    '/lighter.Lightning/CreateInvoice': {
        'entity': 'invoice',
        'action': 'write'
    },
    '/lighter.Lightning/DecodeInvoice': {
        'entity': 'invoice',
        'action': 'read'
    },
    '/lighter.Lightning/GetInfo': {
        'entity': 'info',
        'action': 'read'
    },
    '/lighter.Lightning/ListChannels': {
        'entity': 'channel',
        'action': 'read'
    },
    '/lighter.Lightning/ListInvoices': {
        'entity': 'invoice',
        'action': 'read'
    },
    '/lighter.Lightning/ListPayments': {
        'entity': 'payment',
        'action': 'read'
    },
    '/lighter.Lightning/ListPeers': {
        'entity': 'peer',
        'action': 'read'
    },
    '/lighter.Lightning/ListTransactions': {
        'entity': 'transaction',
        'action': 'read'
    },
    '/lighter.Locker/LockLighter': {
        'entity': 'lock',
        'action': 'write'
    },
    '/lighter.Lightning/NewAddress': {
        'entity': 'address',
        'action': 'write'
    },
    '/lighter.Lightning/OpenChannel': {
        'entity': 'channel',
        'action': 'write'
    },
    '/lighter.Lightning/PayInvoice': {
        'entity': 'payment',
        'action': 'write'
    },
    '/lighter.Lightning/PayOnChain': {
        'entity': 'transaction',
        'action': 'write'
    },
    '/lighter.Lightning/UnlockNode': {
        'entity': 'unlock',
        'action': 'write'
    },
    '/lighter.Lightning/WalletBalance': {
        'entity': 'balance',
        'action': 'read'
    },
}

READ_PERMS = [
    {
        'entity': 'balance',
        'action': 'read'
    },
    {
        'entity': 'channel',
        'action': 'read'
    },
    {
        'entity': 'info',
        'action': 'read'
    },
    {
        'entity': 'invoice',
        'action': 'read'
    },
    {
        'entity': 'payment',
        'action': 'read'
    },
    {
        'entity': 'peer',
        'action': 'read'
    },
    {
        'entity': 'transaction',
        'action': 'read'
    },
]

INVOICE_PERMS = [
    {
        'entity': 'channel',
        'action': 'read'
    },
    {
        'entity': 'info',
        'action': 'read'
    },
    {
        'entity': 'invoice',
        'action': 'read'
    },
    {
        'entity': 'invoice',
        'action': 'write'
    },
    {
        'entity': 'peer',
        'action': 'read'
    },
]
